pika/adapters/blocking_connection.py

```diff
@@ -699,8 +699,10 @@ def process_data_events(self, time_limit=0):
             until I/O produces actionalable events. Defaults to 0 for backward
             compatibility. This parameter is NEW in pika 0.10.0.
         """
-        common_terminator = lambda: bool(
-            self._channels_pending_dispatch or self._ready_events)
+        with self._acquire_event_dispatch() as dispatch_acquired:
+            # Check if we can actually process pending events
+            common_terminator = lambda: bool(dispatch_acquired and
+                (self._channels_pending_dispatch or self._ready_events))
 
         if time_limit is None:
```

Comment:
Right, but then the `yield` exits, hitting the `finally` block, which decrements `_event_dispatch_suspend_depth`. I realize that this is all happening via a single event loop with callbacks. I assume part of the reason for this code is so that later calls don't try to acquire the same resource. In addition, nowhere else in the code is `_acquire_event_dispatch` called and the value used outside of the `yield`.